### ISSUE TITLE
Code lens is not accessible via screen reader

### ISSUE THREAD

**Issue by kimadeline** [microsoft/vscode#171755]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.75.0-insider (Universal) Commit: 5ae522f8563fe48b89470923f77a9732f
- OS Version: macOS Monterey 12.6.2

Steps to Reproduce:

1. Make sure `typescript.referencesCodeLens.enabled` is set to true
2. Enable VoiceOver (System Preferences > Accessibility > VoiceOver)
3. Open a TypeScript file with code lenses, navigate to them
4. Note that they aren't highlighted by the screen reader, nor spoken out loud - I can click on the code lens using my mouse, which will open a side list with the references, and this list will be picked up by the screen reader.

**Comment by meganrogge** [comment]
@alexdima lmk what you think of this, but I would guess we'd check the cursor position and accept the closest one when the keybinding is dispatched

**Comment by alexdima** [comment]
@meganrogge I don't know how we should expose code lenses to screen readers? Do you have any suggestions? Do you know what other editors do?

**Comment by jrieken** [comment]
We have "F1 > Show CodeLens Commands For Current line" which shows (and reads) the lenses as quick pick. I believe that's better than just reading it with source code. However, we might wanna add an audio cue when navigating to a line with lenses

**Comment by meganrogge** [comment]
I discussed this with @jooyoungseo and he suggested that a screen reader user navigating the copilot solutions with arrow keys wants to have a keybinding that will accept the solution. 

In this case, we'd want to run the closest codelens on a line less than or equal to the cursor.  Would that be possible?

**Comment by jrieken** [comment]
Yeah, kinda works like that already today. You invoke the command and pick the lens that you want to execute. The quick pick intermediate steps is there because there can be multiple lenses per line. 

> we'd want to run the closest codelens on a line less than or equal to the cursor

The closest codelens might be hundreds of lines away so I don't know if jumping straight into its lenses is a good idea. How do we communicate where you landed? Maybe it needs a navigate to next/prev line with lens and from there you invoke the current command.

**Comment by meganrogge** [comment]
> Maybe it needs a navigate to next/prev line with lens and from there you invoke the current command.

Is there a way to skip this step?

```
codelens 1

solution 1 line 1
solution 1 line 2

codelens 2

solution 2 line 1 <cursor>
solution 2 line 2

codelens 3

solution 3 line 1
solution 3 line 2
```

`Run current codelens` knows where the code lenses are in the file, knows cursor position, filters for and runs the closest code lens where `codelensLineNumber` <= `cursorLineNumber`

**Comment by jrieken** [comment]
> knows cursor position, filters for and runs the closest code lens where codelensLineNumber <= cursorLineNumber

I understand but my concern is that you can be on line 123456789 and the closest lens is on line 1. That lens might make sense for the current or not - code lenses cannot express that, they only sit with a line, not an associated range. That's why I believe you first need to navigate to the line with a lens, understand its context, and then execute the command of the lens.

**Comment by meganrogge** [comment]
oh, so there are editors that have more than one type of code lens provided... I see what you mean then

**Comment by meganrogge** [comment]
I think the question now is how to communicate the context to the screen reader user so they don't have to arrow key around to figure that out. 

It is still confusing to me that a code lens doesn't have a particular range though.

**Comment by jrieken** [comment]
> It is still confusing to me that a code lens doesn't have a particular range though

Well, that's how we designed them to be and that's how extensions use them. The API does actually spec them to have a range but it says that it must be a single range spec. 

> I think the question now is how to communicate the context to the screen reader user so they don't have to arrow key around to figure that out.

Unfortunately there is no context - lenses can sit with any line and can refer to that exact line (see run command in package.json) or to many lines below (e.g function references, copilot panel etc). We can think about additional API of adding another range that describes the "whole range" a lens applies to but it needs adoption by LSP and other extensions first. 


As far as accessibility goes we already have the "show code lens commands for current line". What's missing is to know if the current line has lenses (audio cue maybe?) and a way to navigate to the next/previous line with a lens.

**Comment by meganrogge** [comment]
> show code lens commands for current line

The issue is when one finds themself in a code block that they want to use, they should be able to accept it immediately without navigating anywhere. 

And they should know this without having to explore - similar to what we do in the diff editor's aria label 

https://user-images.githubusercontent.com/29464607/233150154-64a7965d-6dac-4c57-ae40-cad8cfae16fe.mov

**Comment by jooyoungseo** [comment]
@meganrogge -- As per copilot, it has been now addressed and it is very satisfactory.

> When you open suggestions in a separate tab, this tab supports keyboard navigation: You can use alt+[ and alt+] to navigate to the previous/next suggestion in the panel, and you can use ctrl+/ to accept the selected suggestion.
Retrieved from [community Discussion #7139](https://github.com/orgs/community/discussions/7139).

**Comment by gjsjohnmurray** [comment]
> a way to navigate to the next/previous line with a lens.

:+1:

**Comment by dmitrivMS** [comment]
@copilot try again
